File: leetcode-all/剑指Offer06-从尾到头打印链表.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # 遍历指针进行存储
    def reversePrint(self, head: ListNode):
        ans = []
        while head != None:
            ans.insert(0, head.val)
            head = head.next
        return ans

    # 借助栈，先遍历一遍链表，再反向遍历一遍栈
    def reversePrint_example(self, head: ListNode):
        stack = []
        while head:
            stack.append(head.val)
            head = head.next
        return stack[::-1]

File: 链表/剑指 Offer 06-从尾到头打印链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {number[]}
 */

function ListNode(val) {
    this.val = val;
    this.next = null;
}
 
var reversePrint = function(head) {
    var result = [];
    while(head !== null) {
        result.push(head.val);
        head = head.next;
    }
    var len = result.length - 1, temp;
    for(var i = 0; i < len/2; i++) {
        temp = result[i];
        result[i] = result[len-i];
        result[len-i] = temp;
    }
    return result;
};

var h = new ListNode(1);
var tail = h;
var num = [1, 3, 2];
for (var i = 1; i < num.length; i++){
    var temp = new ListNode(num[i]);
    tail.next = temp;
    tail = tail.next;
}

console.log(reversePrint(h));
